ganttproject/src/main/java/net/sourceforge/ganttproject/storage/SqlProjectDatabaseImpl.kt

```diff
@@ -67,22 +77,33 @@ class SqlProjectDatabaseImpl(private val dataSource: DataSource) : ProjectDataba
     }
   }
 
-  /** Execute query and save its xlog. */
-  private fun withLog(
-    errorMessage: () -> String,
-    buildQuery: (dsl: DSLContext) -> String
-  ): Unit = withDSL(errorMessage) { dsl ->
+  /** Execute queries and save their logs as a transaction. */
+  private fun executeAndLog(queries: List<Query>): Unit = withDSL({ "Failed to commit transaction" }) { dsl ->
     dsl.transaction { config ->
       val context = DSL.using(config)
-      val query = buildQuery(context)
-      context.execute(query)
-      context
-        .insertInto(LOGRECORD)
-        .set(LOGRECORD.SQL_STATEMENT, query)
-        .execute()
+      val nextTxnId = context.nextval(TXNID)
+      queries.forEach {
+        try {
+          context.execute(it.sqlStatement)
+          context
+            .insertInto(LOGRECORD)
+            .set(LOGRECORD.TXN_ID, nextTxnId.toInt())
```

Comment:
It appears that you're mixing the locally generated sequence of numbers with the txn id from the server response. This is not a good idea, because they are different. In the presence of concurrent updates the server chooses the order of committing and the server generates the transaction ids. 

Local ids are also important, they will allow us to establish a reliable order of the xlog records, but they shall not be mixed with the server base txn ids.

Reply:
It seems to be fine as the local ids are never accessed for read, doesn't it? They only specify the order of transactions